### DISCUSSION TITLE
CVODE EEROR

### DISCUSSION THREAD

**Discussion by FernandoIGuerra** [AMReX-Combustion/PeleC#713] [Q&A] [answered]
![WhatsApp Image 2023-11-27 at 12 31 10](https://github.com/AMReX-Combustion/PeleC/assets/102425257/753924ae-7b24-48cf-a60c-011c65d87eb3)
Hi Everyone,
I am simulating a DDT in a obsturcuted channel and the computations fails at some point getting the error message showed in the image. Is there a way to solve it? I tried by increasing the refinement level and reducing the CFL. However, it was not solve. 

Thanks in advance

**Comment by baperry2** [comment]
This type of CVODE error often is a symptom of problems that are unrelated to chemistry integration, but that cause CVODE to crash when it gets passed a bad state. As a first debugging step, I'd recommend dumping a plot file at the time step immediately before it crashes and looking for any issues in the state at that time (unexpectedly high or low temperatures, out of bounds mass fractions, etc), particularly near an EB surface if you have one.

The issue may be related to the chemistry integration process, particularly for large/stiff mechanisms. In that case, adding refinement or reducing the CFL is sometimes sufficient. You can also try using a different solver within CVODE using the `cvode.solve_type` option in your input file (the choices depend on the hardware and compiler that you are using). If you have  `cvode.solve_type = GMRES`, you may also benefit from specifying tighter tolerances for `ode.rtol` and `ode.atol`.

**Reply by FernandoIGuerra** [reply]
Even reducing the CFL, time step of 1e-14, and adding refinement, 1/2048, near at the EB surface, the problem was not solved. I am using the LiDryer chemical mechanism and, near the EB surface, where the nan appears, i obtain pressure around 4MPa due to the formation of Mach Stems. can be the chemical mechanism the problem? If it is so, is it possible to implement a one step Arrhenius chemical mechanism?

I appreciate your help with this matter.

**Comment by drummerdoc** [comment]
Of all the chemistry mechanisms provided with Pele, the LiDryer is probably the most tested and one of the least stiff.  To me, your symptoms really suggest that something is wrong elsewhere.  I would carefully scan the run log information written to stdout, with particular attention on the temperature profiles and near the walls, and also simply plot the plot files leading up to the failure, looking especially near physical and coarse-fine boundaries.  If you don't find anything obvious, you could try forking a copy of PeleC and pushing the complete setup of your case from scratch, then adding a note here on how we can reproduce what you see.  Once you've done this, you may also want to find a different machine to see if you can reproduce the errors on - some times this can be helpful as well.

**Reply by FernandoIGuerra** [reply]
I followed your suggestions and i found that, close to the EB surface, after the shock waves reaches the EB surface, the temperature drops abruptly to 120 K as is shown in the first figure. Additionally, the second image show the minimun temperature over the time. Previously, I added refinment around the EB and was not enough. 

I appreciate your help with this matter.
 
![Captura](https://github.com/AMReX-Combustion/PeleC/assets/102425257/707c0fa4-e30d-42df-b8c6-5f5f83c16c76)
![image](https://github.com/AMReX-Combustion/PeleC/assets/102425257/f16c5201-8a21-444f-87b1-af0448e76c39)

**Reply by marchdf** [reply]
Do you have `pelec.eb_srd_max_order = 0` in your input file? You can also try to get rid of small cells with: `eb2.small_volfrac = 1.e-6` or something like that if you can identify the vfrac of the cell that is throwing such a small temperature.

Other questions:
1. Is this an isothermal wall? 
2. how about your mass fractions? Do you see any of those being out of bounds? You can look here for extra tips: https://amrex-combustion.github.io/PeleC/geometry/EB.html#mass-fractions-at-the-eb. You can also try print out with : `pelec.extrema_spec_name = ALL` to diagnose which mass frac might be out-of-bounds

**Reply by FernandoIGuerra** [reply]
I am still encoutering the error using the CVODE. the error, as was suggested, come up near the EB. I even changed the geometry shape from vertical box to cylinders. The following lines describe smy inputs in the .inp file. I would really appreaciate any help.

eb2.small_volfrac = 1e-06
ebd.boundary_grad_stencil_type = 1
ebd.clean_massfrac = true
ebd_clean_massfrac_threshold = 1.0e-12

pelec.eb_isothermal = 1
pelec.eb_boundary_T = 298.0

pelec.eb_srd_max_order = 0
pelec.extrema_spec_name = ALL
pelec.eb_zero_body_state = true

![image](https://github.com/AMReX-Combustion/PeleC/assets/102425257/7bfffec1-d804-4317-a422-d13fcc2fa794)

**Reply by marchdf** [reply]
These don't look like valid options:
```
ebd.clean_massfrac = true
ebd_clean_massfrac_threshold = 1.0e-12
```
They should be 
```
pelec.eb_clean_massfrac = true
pelec.eb_clean_massfrac_threshold = 1e-12
```
Please look at your log at the end. It will complain about unused input options. A sure sign that you made a mistake defining them.

**Reply by FernandoIGuerra** [reply]
Even with those changes, the Nan values come up around the EB as is show in the previous Figure. Is there any other input that can solve this issue. Ive already refined the mesh size to  1.758e-03 cm.  I would really appreaciate any help.

**Reply by marchdf** [reply]
Try increasing the eb_clean_massfrac_threshold and small_volfrac values. Let us know if that doesn't work.

**Reply by FernandoIGuerra** [reply]
I could simulate some additional times steps. However, the simlation ends with the same error:
From CVODE: At t = 9.07498e-10, mxstep steps taken before reaching tout.
Even, I increased the max number of iteration.
cvode.max_nls_iters = 200000

eb2.small_volfrac = 1.0e-06
ebd.boundary_grad_stencil_type = 1
pelec.eb_clean_massfrac = true
pelec.eb_clean_massfrac_threshold = 1.0e-01

Thanks in advance

**Comment by drummerdoc** [comment, marked as answer]
A couple other tips:  When plotting T, e.g., and searching for weirdness, set the max in the color bar to 298 or something, as it makes it super easy to see low temperature cells.  Also, try to isolate the exact time step where the temperature drops below ambient - ultimately, we're trying to diagnose whether it happens in a chemistry solve, or with the CFD part, and to do this, we need to narrow down which exact operation goes bad.

**Reply by FernandoIGuerra** [reply]
![image](https://github.com/AMReX-Combustion/PeleC/assets/102425257/2fe0a64a-e9f1-4a82-be95-11126d4106ac)
It seems to be the temperature drop due to the negative gradient pressure ahead the obstacle. Even adding a refinement in that region and following the above recommendation the problem wasnot solved. I would be really helpfull if you can take a look on it. I appreciate your help with this matter. 

[DDT_05.zip](https://github.com/AMReX-Combustion/PeleC/files/14107986/DDT_05.zip)

**Reply by pierreadawe-design** [reply]
Hi I know this is old but I am having a similar issue with NaN values forming around an area where there is a transition from deflagration to detonation, did you ever get this resolved?

**Comment by KaisenseChen** [comment]
谢谢你的来信，我会尽快回复你的。

**Reply by pierreadawe-design** [reply]
Thank you, looking forward to your reply!
